rust/azure_iot_operations_connector/src/base_connector/managed_azure_device_registry.rs

```diff
@@ -380,31 +331,144 @@ impl DeviceEndpointClient {
     ///
     /// If the specification mutex has been poisoned, which should not be possible
-    pub async fn recv_update(&mut self) -> Option<()> {
-        // handle the notification
-        // We set auto ack to true, so there's never an ack here to deal with. If we restart, then we'll implicitly
-        // get the update again because we'll pull the latest definition on the restart, so we don't need to get
-        // the notification again.
-        let (updated_device, _) = self.device_update_observation.recv_notification().await?;
-        // update self with updated specification
-        let mut unlocked_specification = self.specification.write().unwrap(); // unwrap can't fail unless lock is poisoned
-        *unlocked_specification = DeviceSpecification::new(
-                updated_device,
-                // TODO: get device_endpoint_credentials_mount_path from connector config
-                "/etc/akri/secrets/device_endpoint_auth",
-                &self.device_endpoint_ref.inbound_endpoint_name,
-            ).expect("Device Update Notification should never provide a device that doesn't have the inbound endpoint");
-        Some(())
+    pub async fn recv_notification(&mut self) -> ClientNotification<AssetClient> {
+        loop {
+            tokio::select! {
+                biased;
+                update = self.device_update_observation.recv_notification() => {
+                    // handle the notification
+                    // We set auto ack to true, so there's never an ack here to deal with. If we restart, then we'll implicitly
+                    // get the update again because we'll pull the latest definition on the restart, so we don't need to get
+                    // the notification again.
+                    let Some((updated_device, _)) = update else {
+                        // if the update notification is None, then the device endpoint has been deleted
+                        // unobserve as cleanup
+                        Self::unobserve_device(&self.connector_context, &self.device_endpoint_ref).await;
+                        return ClientNotification::Deleted;
+                    };
+                    // update self with updated specification
+                    let mut unlocked_specification = self.specification.write().unwrap(); // unwrap can't fail unless lock is poisoned
+                    *unlocked_specification = DeviceSpecification::new(
+                                    updated_device,
+                                    // TODO: get device_endpoint_credentials_mount_path from connector config
+                                    "/etc/akri/secrets/device_endpoint_auth",
+                                    &self.device_endpoint_ref.inbound_endpoint_name,
+                                ).expect("Device Update Notification should never provide a device that doesn't have the inbound endpoint");
```

Comment:
nit: macro is making weird formatting again.

Reply:
ugh, this was driving me crazy in this code - I did another pass, should be better now!